coreneuron/io/setup_fornetcon.cpp

```diff
@@ -0,0 +1,220 @@
+    std::cout << it->first << " type_to_slot " << it->second << std::endl;
+    int mtype = it->first;
+    int fnslot = it->second;
+    int nodecount = nt._ml_list[mtype]->nodecount;
+    std::cout << "nodecount " << nodecount << std::endl;
+    for (int i=0; i < nodecount; ++i) {
+      int* fn = fornetcon_slot(mtype, i, fnslot, nt);
+      std::cout << "i=" << i << " *fn=" << *fn << std::endl;
+    }
+  }
+}
+
+static void print_vector(const char* s, std::vector<size_t>& v) {
+  std::cout << s << " size " << v.size() << std::endl;
+  for (size_t i = 0; i < v.size(); ++i) {
+    std::cout << i << " " << v[i] << std::endl;
+  }
+}
+#endif // for debugging
+
+void setup_fornetcon_info(NrnThread& nt) {
+
+  if (nrn_fornetcon_cnt_ == 0) { return; }
+
+  // Mechanism types in use that have FOR_NETCONS statements
+  // Nice to have the dparam fornetcon slot as well so use map
+  // instead of set
+  std::map<int, int> type_to_slot;
+  for (int i = 0; i < nrn_fornetcon_cnt_; ++i) {
+    int type = nrn_fornetcon_type_[i];
+    Memb_list* ml = nt._ml_list[type];
+    if (ml && ml->nodecount) {
+      type_to_slot.insert(std::pair<int, int>(type, nrn_fornetcon_index_[i]));
+    }
+  }
+  if (type_to_slot.empty()) {
+    return;
+  }
+
+  // How many NetCons (weight groups) are involved.
+  // Also count how many weight groups for each target instance.
+  // For the latter we can count in the dparam fornetcon slot.
+
+  // zero the dparam fornetcon slot for counting and count number of slots.
+  size_t n_perm_indices = 0;
+  for (auto it = type_to_slot.begin(); it != type_to_slot.end(); ++it) {
+    int mtype = it->first;
+    int fnslot = it->second;
+    int nodecount = nt._ml_list[mtype]->nodecount;
+    for (int i=0; i < nodecount; ++i) {
+      int* fn = fornetcon_slot(mtype, i, fnslot, nt);
+      *fn = 0;
+      n_perm_indices += 1;
+    }
+  }
+
+  // Count how many weight groups for each slot and total number of weight groups
+  size_t n_weight_perm = 0;
+  for (int i = 0; i < nt.n_netcon; ++i) {
+    NetCon& nc = nt.netcons[i];
+    int mtype = nc.target_->_type;
+    auto search = type_to_slot.find(mtype);
+    if (search != type_to_slot.end()) {
+      int i_instance = nc.target_->_i_instance;
+      int* fn = fornetcon_slot(mtype, i_instance, search->second, nt);
+      *fn += 1;
+      n_weight_perm += 1;
+    }
+  }
+
+  // Displacement vector has an extra element since the number for last item
+  // at n-1 is x[n] - x[n-1] and number for first is x[0] = 0.
+  nt._fornetcon_perm_indices.resize(n_perm_indices + 1);
+  nt._fornetcon_weight_perm.resize(n_weight_perm);
+
+  // From dparam fornetcon slots, compute displacement vector, and
+  // set the dparam fornetcon slot to the index of the displacement vector
+  // to allow later filling the _fornetcon_weight_perm.
+  size_t i_perm_indices = 0;
+  nt._fornetcon_perm_indices[0] = 0;
+  for (auto it = type_to_slot.begin(); it != type_to_slot.end(); ++it) {
```

Comment:
nitpick: 
  for (const auto& kv : type_to_slot) {
  }

then use : `kv.first` and `kv.second`